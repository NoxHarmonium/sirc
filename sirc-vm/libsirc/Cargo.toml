[package]
name = "libsirc"
version = "0.1.0"
edition = "2024"

[lib]
crate-type = ["lib", "cdylib"]

[dependencies]
toolchain = { path = "../toolchain" }

[build-dependencies]
cbindgen = "0.29.0"
