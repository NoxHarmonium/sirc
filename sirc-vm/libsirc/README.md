# libsirc - C++ Compatible Bindings for the sirc-vm Toolchain

## Summary

This crate exists to expose sirc-vm functionality to external C/C++ projects.

Why? The sirc-tiledit application is a QT C++ application and needs to be able to
do things like export assembly code. So that I don't have to write duplicate
toolchain code in that application, I'm exposing it from this library.

Why didn't I write that application in Rust? To learn about Rust/C++ interop of course
(not for any practical reason).

## Artefacts

This crate, when built with c-build, will generate the following artefacts:

- Dynamic Libraries
- Static Libraries
- Header Files
- pkg-config .pc file

Which is everything we need to package it up and distribute it.

## Versioning

- The major version will be incremented any time there is a breaking change to existing interfaces.
- The minor version will be incremented when there is a new feature.
- The patch version will be incremented when there is a bugfix.

## Usage

### Getting Started

You need to install c-build to use this crate.

```shell
$ cargo install cargo-c@0.10.14+cargo-0.89.0
```

### Building

To simply build everything to make sure everything is working, use the cbuild command.

```shell
$ cargo cbuild
```

This will build the crate in debug mode.

Everything will be in the target folder under the name of your system architecture.
E.g. on my Mac its at `../target/aarch64-apple-darwin/debug/`.

```shell
$ ls -1  ../target/aarch64-apple-darwin/debug/                                                                                      Tue 22 Jul 08:32:59 2025
build/
cargo-c-libsirc.cache
deps/
examples/
include/
incremental/
libsirc-uninstalled.pc
libsirc.a
libsirc.d
libsirc.dylib*
libsirc.h
libsirc.pc
```

### Exporting

#### For sirc-tiledit development

So we don't need to install the libraries into the system directories during development, you can export the files
directly to the sirc-tiledit project.

```shell
$ cargo cinstall --meson-paths --prefix=/ --destdir=../../sirc-tiledit/third-party/libsirc
```

Then when you run `meson setup` it should find the include/library files from that directory.

#### For packaging/system folders

To export the files in the correct file structure ready for packaging, use the cinstall command.

```shell
$ cargo cinstall --meson-paths --prefix=/usr/local/ --destdir=/tmp/blah
```

This will build the crate in release mode.

Everything will be put into the specified destination dir under the correct prefixes and correctly versioned.

Depending on your platform, if you set prefix correctly, setting destdir to `/` will install the libraries in your
system folders so you can link to them without any extra configuration.

```shell
$ find /tmp/blah/                                                                                                                   Tue 22 Jul 08:38:26 2025
/tmp/blah/
/tmp/blah/usr
/tmp/blah/usr/local
/tmp/blah/usr/local/include
/tmp/blah/usr/local/include/libsirc
/tmp/blah/usr/local/include/libsirc/libsirc.h
/tmp/blah/usr/local/lib
/tmp/blah/usr/local/lib/libsirc.1.0.0.dylib
/tmp/blah/usr/local/lib/pkgconfig
/tmp/blah/usr/local/lib/pkgconfig/libsirc.pc
/tmp/blah/usr/local/lib/libsirc.1.dylib
/tmp/blah/usr/local/lib/libsirc.a
/tmp/blah/usr/local/lib/libsirc.dylib
```