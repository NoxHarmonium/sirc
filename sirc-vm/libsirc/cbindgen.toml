# An optional name to use as an include guard
# default: doesn't emit an include guard
include_guard = "libsirc_h"

# Whether to add a `#pragma once` guard
# default: doesn't emit a `#pragma once`
pragma_once = true

# An optional string of text to output between major sections of the generated
# file as a warning against manual editing
#
# default: doesn't emit anything
autogen_warning = "/* Warning, this file is autogenerated by cbindgen. Don't modify this manually. */"

# Whether to include a comment with the version of cbindgen used to generate the file
# default: false
include_version = true

# An optional namespace to output around the generated bindings
# default: doesn't emit a namespace
namespace = "libsirc"
