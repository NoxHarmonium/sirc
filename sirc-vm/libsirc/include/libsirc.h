#ifndef libsirc_h
#define libsirc_h

#pragma once

/* Generated with cbindgen:0.29.0 */

/* Warning, this file is autogenerated by cbindgen. Don't modify this manually. */

#include <cstdarg>
#include <cstdint>
#include <cstdlib>
#include <ostream>
#include <new>

namespace libsirc {

struct CTilemap {
  const uint16_t *packed_pixel_data;
  uintptr_t packed_pixel_data_len;
  uint16_t palette[16];
};

extern "C" {

/// Takes tilemap data and returns a string that contains assembly code that will assemble to the
/// same tilemap.
///
/// Useful when you want to embed tilemap data in a program, which is probably the simplest way to
/// do it at the moment (although it might be possible to just link in a raw binary file).
char *tilemap_to_str(CTilemap tilemap);

/// Frees a string returned by one of the rust functions in this file.
///
/// Should be called when the string is no longer needed to avoid memory leaks.
///
/// # Safety
///
/// The parameter passed to this function _must_ have been allocated by one of the rust functions
/// in this file.
void free_str(char *str);

}  // extern "C"

}  // namespace libsirc

#endif  // libsirc_h
