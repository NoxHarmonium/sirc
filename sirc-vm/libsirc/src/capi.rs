/// This crate isolates FFI functions from the rest of the workspace to keep everything tidy
use std::ffi::{CStr, c_char};
use std::slice;
use toolchain::printers::shared::print_tokens;
use toolchain::types::data::{DataToken, DataType};
use toolchain::types::shared::{LabelToken, NumberToken, NumberType, Token};

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct CTilemap {
    /// The label printed above this tilemap data for humans to read
    pub label: *const c_char,
    /// The comment printed above this tilemap data, so it can be referenced in assembly routines
    pub comment: *const c_char,
    /// Which palette this tilemap uses
    pub palette_index: u16, // Only for reference, doesn't affect the tilemap
    /// The tilemap pixel data packed so there are four 4bpp pixels per word
    pub packed_pixel_data: *const u16,
    /// The length of the packed pixel data in words
    pub packed_pixel_data_len: usize,
}

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct CPalette {
    /// The comment printed above this palette for humans to read
    pub comment: *const c_char,
    /// The 16 entries of the palette.
    pub data: [u16; 16],
}

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct CTilemapExport {
    /// Pointer to the first tilemap
    /// We could have any number of tilemaps (whatever fits in the PPU ram), so we use a pointer
    /// and size combo, so it can be dynamic
    pub tilemaps: *const CTilemap,
    /// The number of tilemaps that have been passed in
    pub tilemaps_len: usize,
    /// The label printed above the palette data, so it can be referenced in assembly routines
    pub palette_label: *const c_char,
    /// The palettes that the tilemaps refer to.
    /// The palette size in the PPU is 16x16 colour palettes (only 16 colours can be addressable at once with 4bpp)
    /// We may as well represent this as a fixed size array to keep it simple
    pub palettes: [CPalette; 16],
}

fn c_str_to_comment_token(x: *const c_char) -> Token {
    if x.is_null() {
        return Token::Comment(String::new());
    }
    Token::Comment(String::from_utf8_lossy(unsafe { CStr::from_ptr(x).to_bytes() }).into_owned())
}

fn u16_to_data_token(x: u16) -> Token {
    Token::Data(DataToken {
        size_bytes: 2,
        value: DataType::Value(NumberToken {
            value: u32::from(x),
            number_type: NumberType::Hex,
        }),
    })
}

fn slice_to_data_tokens(x: &[u16]) -> Vec<Token> {
    x.iter()
        .map(|x| u16_to_data_token(*x))
        .collect::<Vec<Token>>()
}

/// Takes tilemap data and returns a string that contains assembly code that will assemble to the
/// same tilemap.
///
/// Useful when you want to embed tilemap data in a program, which is probably the simplest way to
/// do it at the moment (although it might be possible to just link in a raw binary file).
///
/// # Safety
///
/// The arguments passed to this function should be properly initialised and point to valid memory
/// They will not be modified by this function.
/// The string returned by this function is owned by the rust code and must be freed by the free_str function.
#[unsafe(no_mangle)]
pub unsafe extern "C" fn tilemap_to_str(tilemap_export: CTilemapExport) -> *mut c_char {
    assert!(!tilemap_export.tilemaps.is_null());
    assert!(!tilemap_export.palette_label.is_null());
    let tilemaps =
        unsafe { slice::from_raw_parts(tilemap_export.tilemaps, tilemap_export.tilemaps_len) };
    let tilemap_tokens = tilemaps
        .iter()
        .flat_map(|tilemap| {
            let packed_pixel_data = if tilemap.packed_pixel_data.is_null() {
                &[]
            } else {
                unsafe {
                    slice::from_raw_parts(tilemap.packed_pixel_data, tilemap.packed_pixel_data_len)
                }
            };
            let owned_tilemap_label =
                unsafe { CStr::from_ptr(tilemap.label).to_string_lossy().into_owned() };

            [
                vec![c_str_to_comment_token(tilemap.comment)],
                vec![Token::Label(LabelToken {
                    name: owned_tilemap_label,
                })],
                slice_to_data_tokens(packed_pixel_data),
            ]
            .concat()
        })
        .collect();

    let palette_tokens = tilemap_export
        .palettes
        .iter()
        .filter(|palette| !palette.comment.is_null())
        .flat_map(|palette| {
            [
                vec![c_str_to_comment_token(palette.comment)],
                slice_to_data_tokens(palette.data.as_slice()),
            ]
            .concat()
        })
        .collect();

    let owned_palette_label = unsafe {
        CStr::from_ptr(tilemap_export.palette_label)
            .to_string_lossy()
            .into_owned()
    };
    let all_tokens = [
        tilemap_tokens,
        vec![Token::Label(LabelToken {
            name: owned_palette_label,
        })],
        palette_tokens,
    ]
    .concat();
    let output_string = print_tokens(&all_tokens);

    let c_str = std::ffi::CString::new(output_string).unwrap();
    c_str.into_raw()
}

/// Frees a string returned by one of the rust functions in this file.
///
/// Should be called when the string is no longer needed to avoid memory leaks.
///
/// # Safety
///
/// The parameter passed to this function _must_ have been allocated by one of the rust functions
/// in this file.
#[unsafe(no_mangle)]
pub unsafe extern "C" fn free_str(str: *mut c_char) {
    unsafe {
        let c_str = std::ffi::CString::from_raw(str);
        drop(c_str);
    }
}
