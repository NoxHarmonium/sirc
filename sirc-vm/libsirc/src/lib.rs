/// This crate isolates FFI functions from the rest of the workspace to keep everything tidy
use std::ffi::{c_char, CStr};
use std::slice;
use toolchain::printers::shared::print_tokens;
use toolchain::types::data::{DataToken, DataType};
use toolchain::types::shared::{LabelToken, NumberToken, NumberType, Token};

#[repr(C)]
pub struct CTilemap {
    pub packed_pixel_data: *const u16,
    pub packed_pixel_data_len: usize,
    pub palette: [u16; 16],
}

fn u16_to_data_token(x: u16) -> Token {
    Token::Data(DataToken {
        size_bytes: 2,
        value: DataType::Value(NumberToken {
            value: u32::from(x),
            number_type: NumberType::Hex,
        }),
    })
}

fn slice_to_data_tokens(x: &[u16]) -> Vec<Token> {
    x.iter()
        .map(|x| u16_to_data_token(*x))
        .collect::<Vec<Token>>()
}

/// Takes tilemap data and returns a string that contains assembly code that will assemble to the
/// same tilemap.
///
/// Useful when you want to embed tilemap data in a program, which is probably the simplest way to
/// do it at the moment (although it might be possible to just link in a raw binary file).
///
/// # Safety
///
/// The arguments passed to this function should be properly initialised and point to valid memory
/// They will not be modified by this function.
/// The string returned by this function is owned by the rust code and must be freed by the free_str function.
#[unsafe(no_mangle)]
pub unsafe extern "C" fn tilemap_to_str(label_name: *const c_char, tilemap: CTilemap) -> *mut c_char {
    assert!(!label_name.is_null());
    let owned_label_name = unsafe { CStr::from_ptr(label_name).to_string_lossy().into_owned() };
    let packed_pixel_data = if tilemap.packed_pixel_data.is_null() {
        &[]
    } else {
        unsafe { slice::from_raw_parts(tilemap.packed_pixel_data, tilemap.packed_pixel_data_len) }
    };

    let pixel_tokens = slice_to_data_tokens(packed_pixel_data);
    let palette_tokens = slice_to_data_tokens(&tilemap.palette);
    let all_tokens = [vec![Token::Label(LabelToken { name: owned_label_name })], pixel_tokens, palette_tokens].concat();
    let output_string = print_tokens(&all_tokens);

    let c_str = std::ffi::CString::new(output_string).unwrap();
    c_str.into_raw()
}

/// Frees a string returned by one of the rust functions in this file.
///
/// Should be called when the string is no longer needed to avoid memory leaks.
///
/// # Safety
///
/// The parameter passed to this function _must_ have been allocated by one of the rust functions
/// in this file.
#[unsafe(no_mangle)]
pub unsafe extern "C" fn free_str(str: *mut c_char) {
    unsafe {
        let c_str = std::ffi::CString::from_raw(str);
        drop(c_str);
    }
}
