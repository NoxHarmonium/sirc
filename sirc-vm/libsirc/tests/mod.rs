#![warn(clippy::all, clippy::pedantic, clippy::nursery)]
#![allow(
    // I don't like this rule
    clippy::module_name_repetitions,
    // Not sure what this is, will have to revisit
    clippy::must_use_candidate,
    // Will tackle this at the next clean up
    clippy::too_many_lines,
    // Might be good practice but too much work for now
    clippy::missing_errors_doc,
    // Not stable yet - try again later
    clippy::missing_const_for_fn,
    // I have a lot of temporary panics for debugging that will probably be cleaned up
    clippy::missing_panics_doc
)]
#![deny(warnings)]

use insta::assert_snapshot;
use libsirc::capi::{CPalette, CTilemap, CTilemapExport, free_str, tilemap_to_str};
use std::ffi::CString;

/// Just exists to make sure nothing goes out of scope and invaldates pointers
#[repr(C)]
struct TestHarness {
    palette_comments: [CString; 16],
    tilemap_comment: CString,
    tilemap_label_name: CString,
    palette_label_name: CString,
    palettes: [CPalette; 16],
    tilemap: CTilemap,
    tilemaps: Vec<CTilemap>,
    export: CTilemapExport,
}

fn create_test_export() -> TestHarness {
    let pixels = &[1u16, 2u16, 3u16, 4u16];

    let tilemap_comment = CString::new("tilemap 1 comment").unwrap();
    let palette_label_name = CString::new("palettes").unwrap();
    let tilemap_label_name = CString::new("tilemaps").unwrap();

    let palette_comments: [CString; 16] = (0..16)
        .map(|i| CString::new(format!("palette {i} comment")).unwrap())
        .collect::<Vec<_>>()
        .try_into()
        .unwrap();

    let palettes: [CPalette; 16] = (0..16)
        .map(|i: u16| CPalette {
            comment: palette_comments[i as usize].as_ptr(),
            data: [
                i,
                1 + i,
                2 + i,
                3 + i,
                4 + i,
                5 + i,
                6 + i,
                7 + i,
                8 + i,
                9 + i,
                10 + i,
                11 + i,
                12 + i,
                13 + i,
                14 + i,
                15 + i,
            ],
        })
        .collect::<Vec<_>>()
        .try_into()
        .unwrap();

    let tilemap: CTilemap = CTilemap {
        label: tilemap_label_name.as_ptr(),
        comment: tilemap_comment.as_ptr(),
        packed_pixel_data: pixels.as_ptr(),
        packed_pixel_data_len: pixels.len(),
        palette_index: 0,
    };

    let tilemaps = vec![tilemap];

    let export = CTilemapExport {
        tilemaps: tilemaps.as_ptr(),
        tilemaps_len: 1,
        palette_label: palette_label_name.as_ptr(),
        palettes,
    };

    TestHarness {
        palette_comments,
        tilemap_comment,
        tilemap_label_name,
        palette_label_name,
        palettes,
        tilemap,
        tilemaps,
        export,
    }
}

#[test]
fn test_tilemap_to_str() {
    // Setup
    // Note: This is designed to be called from C/C++ via FFI, this setup code is just simulating that
    let test_harness = create_test_export();
    let export = test_harness.export;

    // Run
    let str = unsafe { tilemap_to_str(export) };

    // Assert
    let c_str = unsafe { CString::from_raw(str) };
    let actual_str = c_str.to_str().unwrap();

    assert_snapshot!(actual_str);

    // No need to free the string in a rust test because it will be freed when c_str goes out of bounds
    // (the from_raw retakes ownership in rust's memory management)
}

#[test]
fn test_free_str() {
    // Setup
    // Note: This is designed to be called from C/C++ via FFI, this setup code is just simulating that\
    let test_harness = create_test_export();
    let export = test_harness.export;

    // Run
    let raw_str = unsafe { tilemap_to_str(export) };
    // We don't use CString::from_raw here because we don't want rust to take ownership of the string
    // again because we specifically want to test the free function.
    unsafe { free_str(raw_str) }

    // Not sure how to assert here. It will panic on double frees or anything like that though
}
