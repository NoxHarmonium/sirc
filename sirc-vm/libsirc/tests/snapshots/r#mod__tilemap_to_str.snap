---
source: libsirc/tests/mod.rs
expression: actual_str
---
;tilemap 1 comment
:tilemaps
.DW #0x0001
.DW #0x0002
.DW #0x0003
.DW #0x0004
:palettes
;palette 0 comment
.DW #0x0000
.DW #0x0001
.DW #0x0002
.DW #0x0003
.DW #0x0004
.DW #0x0005
.DW #0x0006
.DW #0x0007
.DW #0x0008
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
;palette 1 comment
.DW #0x0001
.DW #0x0002
.DW #0x0003
.DW #0x0004
.DW #0x0005
.DW #0x0006
.DW #0x0007
.DW #0x0008
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
;palette 2 comment
.DW #0x0002
.DW #0x0003
.DW #0x0004
.DW #0x0005
.DW #0x0006
.DW #0x0007
.DW #0x0008
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
;palette 3 comment
.DW #0x0003
.DW #0x0004
.DW #0x0005
.DW #0x0006
.DW #0x0007
.DW #0x0008
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
;palette 4 comment
.DW #0x0004
.DW #0x0005
.DW #0x0006
.DW #0x0007
.DW #0x0008
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
;palette 5 comment
.DW #0x0005
.DW #0x0006
.DW #0x0007
.DW #0x0008
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
;palette 6 comment
.DW #0x0006
.DW #0x0007
.DW #0x0008
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
;palette 7 comment
.DW #0x0007
.DW #0x0008
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
.DW #0x0016
;palette 8 comment
.DW #0x0008
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
.DW #0x0016
.DW #0x0017
;palette 9 comment
.DW #0x0009
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
.DW #0x0016
.DW #0x0017
.DW #0x0018
;palette 10 comment
.DW #0x000A
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
.DW #0x0016
.DW #0x0017
.DW #0x0018
.DW #0x0019
;palette 11 comment
.DW #0x000B
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
.DW #0x0016
.DW #0x0017
.DW #0x0018
.DW #0x0019
.DW #0x001A
;palette 12 comment
.DW #0x000C
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
.DW #0x0016
.DW #0x0017
.DW #0x0018
.DW #0x0019
.DW #0x001A
.DW #0x001B
;palette 13 comment
.DW #0x000D
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
.DW #0x0016
.DW #0x0017
.DW #0x0018
.DW #0x0019
.DW #0x001A
.DW #0x001B
.DW #0x001C
;palette 14 comment
.DW #0x000E
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
.DW #0x0016
.DW #0x0017
.DW #0x0018
.DW #0x0019
.DW #0x001A
.DW #0x001B
.DW #0x001C
.DW #0x001D
;palette 15 comment
.DW #0x000F
.DW #0x0010
.DW #0x0011
.DW #0x0012
.DW #0x0013
.DW #0x0014
.DW #0x0015
.DW #0x0016
.DW #0x0017
.DW #0x0018
.DW #0x0019
.DW #0x001A
.DW #0x001B
.DW #0x001C
.DW #0x001D
.DW #0x001E
